pkgs/native_assets_cli/lib/src/model/asset.dart

```diff
@@ -1,219 +1,27 @@
-
-  static const _pathTypeValue = 'absolute';
-
-  @override
-  Map<String, Object> toYaml() => {
-        AssetPath._pathTypeKey: _pathTypeValue,
-        AssetPath._uriKey: uri.toFilePath(),
-      };
-
-  @override
-  int get hashCode => Object.hash(uri, 133711);
-
-  @override
-  bool operator ==(Object other) {
-    if (other is! AssetAbsolutePath) {
-      return false;
-    }
-    return uri == other.uri;
-  }
-}
-
-/// Asset is avaliable on the system `PATH`.
-///
-/// [uri] only contains a file name.
-class AssetSystemPath implements AssetPath, api.AssetSystemPath {
-  @override
-  final Uri uri;
-
-  AssetSystemPath(this.uri);
-
-  static const _pathTypeValue = 'system';
-
-  @override
-  Map<String, Object> toYaml() => {
-        AssetPath._pathTypeKey: _pathTypeValue,
-        AssetPath._uriKey: uri.toFilePath(),
-      };
-
-  @override
-  int get hashCode => Object.hash(uri, 133723);
-
-  @override
-  bool operator ==(Object other) {
-    if (other is! AssetSystemPath) {
-      return false;
-    }
-    return uri == other.uri;
-  }
-}
-
-/// Asset is loaded in the process and symbols are available through
-/// `DynamicLibrary.process()`.
-class AssetInProcess implements AssetPath, api.AssetInProcess {
-  AssetInProcess._();
-
-  static final AssetInProcess _singleton = AssetInProcess._();
-
-  factory AssetInProcess() => _singleton;
-
-  static const _pathTypeValue = 'process';
-
-  @override
-  Map<String, Object> toYaml() => {
-        AssetPath._pathTypeKey: _pathTypeValue,
-      };
-}
-
-/// Asset is embedded in executable and symbols are available through
-/// `DynamicLibrary.executable()`.
-class AssetInExecutable implements AssetPath, api.AssetInExecutable {
-  AssetInExecutable._();
-
-  static final AssetInExecutable _singleton = AssetInExecutable._();
-
-  factory AssetInExecutable() => _singleton;
-
-  static const _pathTypeValue = 'executable';
-
-  @override
-  Map<String, Object> toYaml() => {
-        AssetPath._pathTypeKey: _pathTypeValue,
-      };
-}
-
-class Asset implements api.Asset {
-  @override
-  final LinkMode linkMode;
-  @override
-  final String id;
-  @override
-  final Target target;
-  @override
-  final AssetPath path;
-
-  Asset({
-    required this.id,
-    required this.linkMode,
-    required this.target,
-    required this.path,
-  });
-
-  factory Asset.fromYaml(YamlMap yamlMap) => Asset(
-        id: as<String>(yamlMap[_idKey]),
-        path: AssetPath.fromYaml(as<YamlMap>(yamlMap[_pathKey])),
-        target: Target.fromString(as<String>(yamlMap[_targetKey])),
-        linkMode: LinkMode.fromName(as<String>(yamlMap[_linkModeKey])),
-      );
-
-  static List<Asset> listFromYamlString(String yaml) {
-    final yamlObject = loadYaml(yaml);
-    if (yamlObject == null) {
-      return [];
+part of '../api/asset.dart';
+
+abstract final class AssetImpl implements Asset {
+  Map<String, Object> toYaml(Version version);
+
+  static List<AssetImpl> listFromYamlList(YamlList list) {
```

Comment:
Is there any **really good reason** we need yaml and not json? If not, I suggest to make this a json protocol.

Yaml mostly makes sense for hand-written things, but neither build config nor build output is hand written. And if there's some use cases for testing, one can always put a yaml2json step in front of passing it to the `build.dart` script.

Reply:
Tracked in https://github.com/dart-lang/native/issues/991. Will do in a follow up PR.